### PR TITLE
Datadog tracing extension: remove OpenTracing

### PR DESCRIPTION
**Pull request by dgoffredo**
Commit Message: Datadog tracing extension uses new tracing library
Additional Description: See below.
Risk Level: Medium
Testing: Existing unit tests are sufficient.
Docs Changes: User-facing documentation for the Datadog tracing integration does not need to change, because these changes do not change the configuration.
Release Notes:  N/A or unsure
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue] [#21083](https://github.com/envoyproxy/envoy/issues/21083)
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

Overview
--------
This is rewrite of the Datadog tracing extension.

Now that there is a non-OpenTracing Datadog C++ Tracing library
([dd-trace-cpp][1]), the first order of business is to remove the
OpenTracing dependency on Envoy's Datadog tracing extension by rewriting
the extension in terms of the new library.

This pull request addresses a longstanding [issue][2], and if accepted will
allow Datadog tracing in Envoy to catch up with Datadog tracing in other
projects. Due to the OpenTracing dependency of the existing extension, Datadog
tracing has been frozen on an [old version][3] for a long time.

Changing to a non-OpenTracing, recent tracing library allows us to address
multiple problems with the current extension:

- improper handling of `error.*` span tags,
- lack of support for Datadog-specific sampling controls such as the
  `DD_TRACE_SAMPLE_RATE`, `DD_TRACE_RATE_LIMIT`, and `DD_SPAN_SAMPLING_RULES`
  environment variables,
- malformed traces resulting from multiple request passes through a proxy within
  the same trace.

In general, we have not been able to update the Datadog tracing extension since
January 2021. With these changes we will be able to make updates from now on.

Design
------
The components of this extension are described in the included `README.md` file.
There's even a diagram.

There is further description of the inner workings of the underlying tracing
library in its own [readme][4] and more extensively in its [maintainer
documentation][5].

[dd-trace-cpp][1] exports mostly concrete types. However, there are two
context-dependent interfaces that are implemented in this pull request.
See `event_scheduler.{h,cc}` and `agent_http_client.{h,cc}`.

Testing
-------
I made minimal modifications to the existing unit tests, and added several new ones.

There is also a `demo/` directory in the source tree that I used for manual
ad hoc integration testing during development. If that directory does not
belong, then we can remove it.

Code Style
----------
I read Envoy's contributing guidelines and C++ style guide, and have formatted
the code using `ci/run_envoy_docker.sh './ci/do_ci.sh format'`.

This is my first time contributing to Envoy. Anywhere that the style of this
extension differs from Envoy's house style, I would be happy to make changes
to conform to the house style.

Datadog Contacts
----------------
~~I'll be on vacation for three weeks right after I propose this pull request.
Until I return (or succumb to the temptation to check this pull request),
my teammate @cgilmour can address questions and make modifications.~~

We're willing to make modifications to [dd-trace-cpp][1] per this code review
as well. Envoy will be the first client of the new library.

Let's get this code in ship shape so that we can carry on adding features.

[1]: https://github.com/DataDog/dd-trace-cpp
[2]: https://github.com/envoyproxy/envoy/issues/21083
[3]: https://github.com/envoyproxy/envoy/issues/23358
[4]: https://github.com/DataDog/dd-trace-cpp/blob/main/README.md
[5]: https://github.com/DataDog/dd-trace-cpp/tree/main/doc

### CODE REVIEW

**Comment by cgilmour** [comment]
As David is away for a bit, I'll sort out the conflicts and DCO so this can at least have CI checks run on them.

File: bazel/repositories.bzl
Comment (moderation) [inline bazel/repositories.bzl:571]: The `msgpack` dependency is only used by the Datadog extension being replaced and can therefore be removed too:

https://github.com/envoyproxy/envoy/blob/main/bazel/repository_locations.bzl#L732-L746

**Review by moderation — CHANGES_REQUESTED** [review]

**Comment by dgoffredo** [comment]
I'm back.

The format checker does not like our use of `std::optional` and `std::string_view`.

https://dev.azure.com/cncf/envoy/_build/results?buildId=122576&view=logs&j=55bc2ab9-36c1-5ab8-58a6-8c258728eeff&t=33b57bf9-ef13-58c8-bc2b-f804aeb16d39

I looked through the history of this policy, and it looks like forbidding the use of the `std` types has to do with supporting iOS 11. Correct me if I'm wrong.

In any case, I'm looking into adding an alternative build mode to `dd-trace-cpp` that will allow us to use Abseil types in the interface, rather than standard types.

Otherwise this PR remains ready for review.

**Comment by dgoffredo** [comment]
If I'm reading https://dev.azure.com/cncf/envoy/_build/results?buildId=122708&view=results correctly, the CI checks have the following complaints:

1. `override` is missing from the destructor of a derived class whose base class has a `virtual` destructor.
2. Some of the function names `look_like_this` instead of `lookLikeThis`.
3. The unit test code coverage for `source/extensions/tracers/datadog` is lower than the allowed minimum of 96.6% (at 65.4%).
4. The `verify_x64 > Verify packages` check failed with:
```
PackagesDistroChecker ERROR [distros] [debian_bullseye/envoy-1.25:proxy-responds] Test failed
stdout: ERROR Envoy did not respond correctly
Response was
Error - Request ID: 01GKMF7PYDY48GJ95D59QAF8VA

Log:
Envoy did not respond correctly
Response was
Error - Request ID: 01GKMF7PYDY48GJ95D59QAF8VA

Log:
```

(1) I'm happy to fix, though it seems unnecessary.
(2) I'm happy to fix.
(3) I will look to address.
(4) I don't understand.

**Comment by dgoffredo** [comment]
I'm currently working on increasing the code coverage of the Datadog tracing unit tests.

Don't hesitate to submit review comments or concerns in the mean time. This integration is my top priority.

**Comment by dgoffredo** [comment]
It looks like I have appeased the CI system. :heavy_check_mark:

**Comment by RyanTheOptimist** [comment]
@cgilmour can you take another look at this PR?

**Comment by dgoffredo** [comment]
@cgilmour is on holiday as of today, I believe.

He has already looked over the code, though he did not submit a review.

I'm interested in getting feedback from an Envoy maintainer, if one is available to do a code review.

**Comment by RyanTheOptimist** [comment]
@dgoffredo looks like you need to fix DCO: https://github.com/envoyproxy/envoy/pull/23958/checks?check_run_id=10076026264

**Comment by dgoffredo** [comment]
Oops, fixed.

**Comment by dgoffredo** [comment]
Ping, @tonya11en. Happy holidays.

File: source/extensions/tracers/datadog/span.cc
Comment (dgoffredo) [inline source/extensions/tracers/datadog/span.cc:90]: Note to self: Check that this isn't a bug. We want the "100% sampling" setting of, e.g., Istio, to make this tracer use "priority sampling," rather than this hard override.

**Comment by dgoffredo** [comment]
Ping, @tonya11en. Happy New Year.

**Comment by dgoffredo** [comment]
Happy Monday, @RyanTheOptimist. Is there anything else I need to do to prepare this PR for review?

**Comment by RyanTheOptimist** [comment]
Hm. Sorry for the delay! I'm the dependency-shepherd for this PR so my work is mostly done :) but I'll ping @tonya11en (the maintainer reviewer) offline and see if that moves the ball forward.

**Comment by RyanTheOptimist** [comment]
@jmarantz, can you please review? @tonya11en seems to be away.

**Comment by KBaichoo** [comment]
/wait

For merge conflicts.

@jmarantz I think this is waiting on your input

**Comment by dgoffredo** [comment]
Hi, @KBaichoo. Which merge conflicts are you referring to?

**Comment by dgoffredo** [comment]
The merge conflict is resolved.

I don't know why it says I requested review from all of those people. I did not.

In trying to get the signoff to work for a couple of commits I made, I ended up doing a `git rebase --signoff <COMMIT>`. That added my signoff signature to other people's commits, which is bad. Ideally I would have added my signoff only to my two commits. However, between those two commits was a merge, and I could not get git to show both commits in a `git rebase -i`.

Perhaps I should squash all of this PR's commits into one on top of the current `main`. I'll see if I can do that.

**Comment by dgoffredo** [comment]
There we go. All changes have been squashed into one commit, DCO is passing, and merge conflicts are resolved.

**Comment by jmarantz** [comment]
and if you need to merge main in the future do so as suggested in https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#submitting-a-pr 👎 
```
branch=$(git status|head -1|cut -f3 -d\ )
git checkout main
git pull
git checkout "$branch"
git pull
```

**Comment by dgoffredo** [comment]
Sorry about that, @jmarantz, and thank you for the links.

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.cc:41]: `const std::string& body`

I'm not sure about the other by-value objects.

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.cc:46]: suggest a temp for `message->headers()`

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.cc:102]: use a concrete type for this `auto`, as the type is not obvious from context.

https://google.github.io/styleguide/cppguide.html#Type_deduction

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.cc:119]: use a concrete type for this one too.

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.cc:121]: combine these two lines

File: source/extensions/tracers/datadog/agent_http_client.h
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.h:31]: we typically put the private member vars at the end of the class definition

https://google.github.io/styleguide/cppguide.html#Class_Format

File: source/extensions/tracers/datadog/agent_http_client.h
Comment (jmarantz) [inline source/extensions/tracers/datadog/agent_http_client.h:36]: use a reference here, or if this is optional, then an OptRef.

File: source/extensions/tracers/datadog/dd.h
Comment (jmarantz) [inline source/extensions/tracers/datadog/dd.h:12]: let's put this using declaration in the .cc files that  need it.

**Review by jmarantz — COMMENTED** [review]
flushing comments ; this is all trivial style stuff so far.

**Review by jmarantz — COMMENTED** [review]
At a higher level, I wonder if you'd mind breaking this PR into a few architectural levels, and we can get them in one at a time.

WDYT?

It's a lot of new code to inhale in one shot especially as I don't really know Datadog or the tracing system, and I'd like to be able to understand it in smaller chunks.

/wait

File: source/extensions/tracers/datadog/README.md
Comment (dgoffredo) [reply source/extensions/tracers/datadog/README.md:27]: It's generated from `diagram.dot` per the comments at the top.

I just noticed that the `.svg` is excluded by a toplevel `.gitignore` rule:
```
david@ein:~/src/envoy$ git check-ignore --verbose source/extensions/tracers/datadog/diagram.svg
.gitignore:40:*.s??     source/extensions/tracers/datadog/diagram.svg
david@ein:~/src/envoy$ fgrep -C 5 '*.s??' .gitignore
/docs/landing_source/.bundle
/generated
*.pyc
**/pyformat
SOURCE_VERSION
*.s??
tags
TAGS
/test/coverage/BUILD
/tools/spelling/.aspell.en.pws
clang-tidy-fixes.yaml
david@ein:~/src/envoy$ 
```

It must have gotten lost in a rebase. I'll add it back.

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.cc:41]: This function is an override of `virtual ... datadog::tracing::HTTPClient::post(...)`: https://github.com/DataDog/dd-trace-cpp/blob/cdb5f235103f7733cb5b8353ac565be42e833af6/src/datadog/http_client.h#L40-L49

The signatures must match, so we can't change it here without also changing it everywhere else (in the base library and in all clients of the base library... which isn't all that many clients right now).

I chose `std::string body` instead of `const std::string& body` so that the value can be `move`d from its original source if possible.

As for the `std::function` callbacks, I'm not sure why I passed those by value and not by `const&`. I suppose the same `move` argument applies, but for the callbacks it's unlikely to matter.

**Comment by dgoffredo** [comment]
Thanks for the first review pass, Joshua.

> At a higher level, I wonder if you'd mind breaking this PR into a few architectural levels, and we can get them in one at a time.
>
> WDYT?
>
> It's a lot of new code to inhale in one shot especially as I don't really know Datadog or the tracing system, and I'd like to be able to understand it in smaller chunks.

Do you mean break it up into a series of different pull requests? If so, I don't know how to do that. There is no incremental path from tracing using the OpenTracing-based adapter to using this new library (dd-trace-cpp) instead.

What might be easier is to review the pull request one component (translation unit) at a time. You've already made some comments on the HTTP client, so we could for example do another pass to dive deep into that part of the tracer.

Then we can continue with each component in turn until you're happy with the whole thing. What do you think?

File: source/extensions/tracers/datadog/README.md
Comment (dgoffredo) [reply source/extensions/tracers/datadog/README.md]: You're right, spelling out the namespace is not much of a problem. Fixed.

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.cc]: done

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.cc]: done

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.cc]: :heavy_check_mark:

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.cc]: :heavy_check_mark:

File: source/extensions/tracers/datadog/agent_http_client.h
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.h]: Ok, I moved the private members of all classes in header files.

File: source/extensions/tracers/datadog/agent_http_client.h
Comment (dgoffredo) [reply source/extensions/tracers/datadog/agent_http_client.h]: :heavy_check_mark:

File: source/extensions/tracers/datadog/dd.h
Comment (dgoffredo) [reply source/extensions/tracers/datadog/dd.h]: I removed the alias. It's unnecessary.

File: source/extensions/tracers/datadog/agent_http_client.cc
Comment (jmarantz) [reply source/extensions/tracers/datadog/agent_http_client.cc]: Cool. I had to do some experiments to see if that usage model actually saved the copy; it does appear to.

**Comment by jmarantz** [comment]
What I was thinking was that each "layer" in your system would have its own unit tests so we could tackle those bottom up, even though during the interim the system wouldn't be functional yet.

**Comment by dgoffredo** [comment]
I'd prefer not to break this pull request into multiple pull requests. If you require it, then I will of course oblige, but here are my thoughts against:

- Each PR will check in dead code. Not "functional yet," as you say, but I don't like the idea of checking in code that does not affect the behavior of the program until future code is also checked in.
- The coverage of the monolithic unit test is good. Breaking it up into per-component tests might be better for organization, but most of the tests are trivial and I think it's better to keep them in the same place.
- Incremental review can still proceed in this pull request. I don't think the separated diffs much facilitates review.
- It'd be some more development work to produce a build of Envoy that uses the old code but also depends on the new code (but doesn't use it).

As for the "layers" you mentioned, here's a dependency graph of the components proposed in this pull request:

g

There appear to be four layers of dependency, but really that's more an artifact of how I chose to organize the code. Here's a synopsis from a review perspective:

- `time_util`, `string_util`, `dict_util`, `tracer_stats`, and `logger` each contain tiny amounts of code.
- Two non-trivial components are `agent_http_client` and `event_scheduler`.
- - `span` just forwards its methods to a data member, and has no real behavior.
- `tracer` is the most important component
- `config` is just a factory for `tracer`.

Let me know how we should proceed.

**Comment by jmarantz** [comment]
Here's my thinking on your points:
 * Each PR will check in dead code. Not "functional yet," as you say, but I don't like the idea of checking in code that does not affect the behavior of the program until future code is also checked in.  **--> think it's fine to check in tested subsystems that aren't live in production yet; I have done this many times in Envoy**
 * The coverage of the monolithic unit test is good. Breaking it up into per-component tests might be better for organization, but most of the tests are trivial and I think it's better to keep them in the same place. **-->I would prefer to see unit tests for each subsystem**
 * Incremental review can still proceed in this pull request. I don't think the separated diffs much facilitate review. **--> it's possible to do this but I'd prefer to have individual reviews of each subsystem and get these checked in**
 * It'd be some more development work to produce a build of Envoy that uses the old code but also depends on the new code (but doesn't use it) **--> I'm not following why that would be needed. I think you can put the new system in place with its unit tests, and just not make it live until you make a single, tiny final change to flip the switch to default to the new implementation.**

The above graph is nice. I'd love to review separate tiny PRs for each of
 * time_util
 * string_util
 * dict_util
 * tracer_stats

and then move onto the next level in your graph. By their names I wonder if the first two of those should be merged into existing similar functionality in Envoy (are DD's string and time needs that different from the rest of Envoy)?

**Comment by jmarantz** [comment]
Another point is that there's a bunch of stuff in this PR that I know nothing about; stuff relating to dependencies and other things, that I'd send off to others to review; I don't want to learn enough to intelligently review all of this PR.

**Comment by dgoffredo** [comment]
Very well, I'll start breaking down the components into separate PRs and mention them here as they become ready.

**Comment by dgoffredo** [comment]
Part 2/9 is grinding through CI: https://github.com/envoyproxy/envoy/pull/25020

**Comment by dgoffredo** [comment]
Part 3/9 is grinding through CI: https://github.com/envoyproxy/envoy/pull/25021

**Comment by dgoffredo** [comment]
Part 4/9 is grinding through CI: https://github.com/envoyproxy/envoy/pull/25025

**Comment by jmarantz** [comment]
FYI we might want to get through the first few first, rather than queuing up all of them. Changes in the lower level ones may impact the ones that depend on them.

**Comment by jmarantz** [comment]
@kyessenov FYI -- Kuat can help me review some of these :) Thanks Kuat!

**Comment by jmarantz** [comment]
@wbpcode thanks for taking a look. @dgoffredo has already broken this up into some smaller PRs -- they are broken out in comments above. Feel free to grab some of them; I'll flip this one into draft mode to avoid confusion.

**Comment by dgoffredo** [comment]
Part 5/9 is grinding through the CI: https://github.com/envoyproxy/envoy/pull/25121

**Comment by dgoffredo** [comment]
Part 6/9 is grinding through the CI: https://github.com/envoyproxy/envoy/pull/25170

**Comment by wbpcode** [comment]
@dgoffredo thanks for these great contributions.
This is a super big prs and it's Chinese New year.
So, I can only review these prs in the next week. Sorry for the delay.

Thanks again.

**Comment by dgoffredo** [comment]
@wbpcode Happy New year! Take your time, I have plenty more to do before I'm blocked on these PRs.

**Comment by dgoffredo** [comment]
Part 7/9 is grinding through the CI: https://github.com/envoyproxy/envoy/pull/25417

**Comment by dgoffredo** [comment]
Here's an update on the status of these pull requests.

I'll propose two or three more pull requests.

"8/9" will introduce the `Tracer` class (as is seen in this original pull request) and its unit tests.

"9/9" will introduce the `Config` class (as is seen in this original pull request) and its unit tests.

"10/9" will remove the old implementation, and probably update the version of the dd-trace-cpp dependency (it's changed since work began on this within Envoy).

It's possible that "10/9" will be part of "9/9" instead.

This week is "R&D Week" at my company, so I'm working on other things. I'll resume work on the remaining pull requests next week.

**Comment by dgoffredo** [comment]
Part 8/9 is grinding through the CI: https://github.com/envoyproxy/envoy/pull/26148

**Comment by dgoffredo** [comment]
Part 9/9 is grinding through the CI: https://github.com/envoyproxy/envoy/pull/26284

Rejoice!

**Comment by dgoffredo** [comment]
/wait

I'll close this (without merging) once https://github.com/envoyproxy/envoy/pull/26284 is merged.

**Comment by dgoffredo** [comment]
https://github.com/envoyproxy/envoy/pull/26284 is merged, so this work is now complete. Thanks for playing.

**Pull request closed without merging**
